File: java_Gemini_Pro/State/Round3/Code.java
package State.Round3;

public class Code {
    public static void main(String[] args) {
        Light light = new Light();
        System.out.println("Initial state: " + light.getState());

        light.turnOn();
        System.out.println("State after turning on: " + light.getState());

        light.turnOn();
        System.out.println("State after trying to turn on again: " + light.getState());

        light.turnOff();
        System.out.println("State after turning off: " + light.getState());

        light.turnOff();
        System.out.println("State after trying to turn off again: " + light.getState());
    }
}

class Light {
    private State state;

    public Light() {
        this.state = new OffState();
    }

    public void setState(State state) {
        this.state = state;
    }

    public void turnOn() {
        state.turnOn(this);
    }

    public void turnOff() {
        state.turnOff(this);
    }

    public String getState() {
        return state.toString();
    }
}

interface State {
    void turnOn(Light light);
    void turnOff(Light light);
}

class OnState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("Light is already on.");
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Turning light off.");
        light.setState(new OffState());
    }

    @Override
    public String toString() {
        return "On";
    }
}

class OffState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("Turning light on.");
        light.setState(new OnState());
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Light is already off.");
    }

    @Override
    public String toString() {
        return "Off";
    }
}

File: Python_Gemini_Flash/State/Round3/Code.py
from abc import ABC, abstractmethod

class LightState(ABC):
  """Abstract base class representing the light state"""

  @abstractmethod
  def turn_on(self, light: "Light") -> "LightState":
    pass

  @abstractmethod
  def turn_off(self, light: "Light") -> "LightState":
    pass

class OnState(LightState):
  """Concrete state representing the light being on"""

  def turn_on(self, light):
    print("Light is already on.")
    return self

  def turn_off(self, light):
    print("Turning light off.")
    return OffState()

class OffState(LightState):
  """Concrete state representing the light being off"""

  def turn_on(self, light):
    print("Turning light on.")
    return OnState()

  def turn_off(self, light):
    print("Light is already off.")
    return self

class Light:
  """Light class that manages its state"""

  def __init__(self):
    self.state = OffState()  # Initial state is Off

  def turn_on(self):
    """Delegates turn on action to the current state"""
    self.state = self.state.turn_on(self)

  def turn_off(self):
    """Delegates turn off action to the current state"""
    self.state = self.state.turn_off(self)

# Example Usage
light = Light()

light.turn_on()  # Output: Turning light on.
light.turn_on()  # Output: Light is already on.
light.turn_off()  # Output: Turning light off.
light.turn_off()  # Output: Light is already off.
